### PR TITLE
MR-733 - Allow address edit of assets with no UPRN 

### PR DESCRIPTION
**Pull request by ange47rm**
- Asset View now displays "Edit address details" button regardless of whether the asset has a UPRN or not
- If the asset has a UPRN, nothing changes, usual behaviour.
- If the asset does not have a UPRN, only one set of editable fields are shown in the "Edit address" page. These are pre-filled with current address details (as in this case no LLPG request is fired).

**Changed behaviour when the LLPG request fails**
When we entered the "Edit address" page previously, for an asset with a UPRN, if the LLPG request failed, we would display the current address read-only fields (on the right) and on the left, we'd display all editable address fields, but blank. This behaviour has been changed to be similar to the non-UPRN edit, meaning that if the LLPG request fails, the read-only "current address" on the right is no longer displayed. Only one set of editable fields is shown, and these, instead of being blank, are pre-populated with the current address's details. The screen would look like "Address without UPRN" screenshot below, with the addition "couldn't fetch LLPG address - manual edit enabled" error on top.

Address without UPRN - Existing address details are loaded into the editable fields.
image

Address with UPRN (unchanged):
image

### CODE REVIEW

File: src/views/asset-edit-view/layout.tsx [hunk 1]
```diff
@@ -32,33 +32,40 @@ export const AssetEditLayout = ({
...
+            setLlpgAddress(searchAddressResponse.addresses[0]);
+          }
+        })
+        .catch(() => {
+          setErrorHeading(
+            "Unable to retrieve address suggestion from the Local Gazetteer",
+          );
+          setErrorDescription(
+            "Please refresh the page and try again, otherwise you are still able to edit the blank fields manually.",
+          );
```
Comment (LBHCallumM) [inline src/views/asset-edit-view/layout.tsx:49]: This error is a bit wordy. How about just using the fist part: "Please refresh the page and try again"

File: src/views/asset-edit-view/__snapshots__/asset-edit-view.test.tsx.snap [hunk 2]
```diff
@@ -16,7 +16,7 @@ exports[`renders the whole 'Edit property address' view 1`] = `
   <span
     class="govuk-caption-m lbh-caption"
   >
-    New Addresses are suggested from the Local Gazetteer to bring some standardisation.
+    Addresses are suggested from the Local Gazetteer to bring some standardisation.
```
Comment (LBHCallumM) [inline src/views/asset-edit-view/__snapshots__/asset-edit-view.test.tsx.snap:19]: Is there a reason for 'New' being removed?

File: src/components/edit-asset-address-form/editable-address/utils.ts [hunk 3]
```diff
@@ -73,3 +74,25 @@ export const buildAssetAddress = (
...
+export const getCurrentAddressFormValues = (editableAddress: AssetAddress) => {
+  return {
+    postPreamble: editableAddress?.postPreamble ? editableAddress?.postPreamble : "",
+    addressLine1: editableAddress?.addressLine1 ? editableAddress.addressLine1 : "",
+    addressLine2: editableAddress?.addressLine2 ? editableAddress.addressLine2 : "",
+    addressLine3: editableAddress?.addressLine3 ? editableAddress.addressLine3 : "",
+    addressLine4: editableAddress?.addressLine4 ? editableAddress.addressLine4 : "",
+    postcode: editableAddress?.postCode ? editableAddress.postCode : "",
+  };
+};
```
Comment (LBHCallumM) [inline src/components/edit-asset-address-form/editable-address/utils.ts:98]: Dont both of these methods have the same functionality?

File: src/components/edit-asset-address-form/editable-address/editable-address.tsx [hunk 4]
```diff
@@ -134,7 +102,14 @@ export const EditableAddress = ({
       });
   };
 
-  if (!llpgAddress && loading) {
+  const getFormInitialValues = () => {
+    if (llpgAddress && assetHasUprn) {
+      return getLlpgAddressFormValues(llpgAddress);
+    }
+    return getCurrentAddressFormValues(currentAddress);
+  };
```
Comment (LBHCallumM) [inline src/components/edit-asset-address-form/editable-address/editable-address.tsx:110]: Refer to the previous comment. These methods do the same thing

**Review by LBHCallumM — APPROVED** [review]
Added a few comments

File: src/views/asset-edit-view/layout.tsx (hunk 1 above)
Comment (ange47rm) [reply src/views/asset-edit-view/layout.tsx:49]: This was suggested by Sean P a while back, so I'm tempted to leave it as it is

File: src/views/asset-edit-view/__snapshots__/asset-edit-view.test.tsx.snap (hunk 2 above)
Comment (ange47rm) [reply src/views/asset-edit-view/__snapshots__/asset-edit-view.test.tsx.snap:19]: I read it and thought it shouldn't be there. They're not "new" addresses, they're just suggestions for existing addresses. If you disagree I'm happy to leave it as is though

File: src/components/edit-asset-address-form/editable-address/utils.ts (hunk 3 above)
Comment (ange47rm) [reply src/components/edit-asset-address-form/editable-address/utils.ts:98]: Similar, I tried to see if typeOf could be used but in this case it couldn't, as it only works with JS types (strings, number, etc) and not custom ones/interfaces.

The difference is that they take two different parameters. One takes in an Address (which would be an LLPG suggestion) and one an AssetAddress (from our DynamoDb). The property names are a little different

File: src/components/edit-asset-address-form/editable-address/editable-address.tsx (hunk 4 above)
Comment (ange47rm) [reply src/components/edit-asset-address-form/editable-address/editable-address.tsx:110]: They work similarly but don't do the same thing, they take parameters of different types

File: src/components/edit-asset-address-form/editable-address/utils.ts (hunk 3 above)
Comment (LBHCallumM) [reply src/components/edit-asset-address-form/editable-address/utils.ts:98]: Could you use an or operator? for example `(editableAddress: AssetAddress | Address) =>`

Comment (LBHCallumM) [reply src/components/edit-asset-address-form/editable-address/utils.ts:98]: Alternatively, make the method more generic, and convert the types into that generic type

Comment (ange47rm) [reply src/components/edit-asset-address-form/editable-address/utils.ts:98]: Initially I tried to use (editableAddress: AssetAddress | Address) and I was trying to couple it with a typeOf to determine the type of the input parameter but I spent some time on it and got nowhere.

Only thing I can think of is to create a new type of object that the "unified" method will take, and then create a new instance of this object using the Address object and AssetAddress object. This way we have one method, one input type, but I just think it's unnecessary complexity.

Did you have anything in mind in particular? I'm out of ideas

Comment (ange47rm) [reply src/components/edit-asset-address-form/editable-address/utils.ts]: Changed to

```
export const getLlpgAddressFormValues = (llpgAddress: Address) => {
  return {
    postPreamble: "",
    addressLine1: llpgAddress?.line1 || "",
    addressLine2: llpgAddress?.line2 || "",
    addressLine3: llpgAddress?.line3 || "",
    addressLine4: llpgAddress?.town || "",
    postcode: llpgAddress?.postcode || "",
  };
};

export const getCurrentAddressFormValues = (currentAddress: AssetAddress) => {
  return {
    postPreamble: currentAddress?.postPreamble || "",
    addressLine1: currentAddress?.addressLine1 || "",
    addressLine2: currentAddress?.addressLine2 || "",
    addressLine3: currentAddress?.addressLine3 || "",
    addressLine4: currentAddress?.addressLine4 || "",
    postcode: currentAddress?.postCode || "",
  };
};
```

as discussed
